modules/core/src/main/scala/fs2/kafka/internal/WithProducer.scala

```diff
@@ -35,24 +36,28 @@ private[kafka] object WithProducer {
   def apply[F[_], K, V](
     settings: TransactionalProducerSettings[F, K, V]
   )(
-    implicit F: Sync[F],
+    implicit F: Concurrent[F],
     context: ContextShift[F]
   ): Resource[F, WithProducer[F]] =
-    blockingResource(settings.producerSettings).flatMap { blocking =>
-      Resource[F, WithProducer[F]] {
-        settings.producerSettings.createProducer.flatMap { producer =>
-          val withProducer = create(producer, blocking)
+    (blockingResource(settings.producerSettings), Resource.eval(Semaphore(1))).tupled.flatMap {
+      case (_blocking, transactionSemaphore) =>
+        Resource[F, WithProducer[F]] {
+          settings.producerSettings.createProducer.flatMap { producer =>
+            val withProducer = new WithProducer[F] {
+              override def apply[A](f: (KafkaByteProducer, Blocking[F]) => F[A]): F[A] =
+                transactionSemaphore.withPermit(f(producer, _blocking))
```

Comment:
@bplommer Is this what you meant when you said to put the semaphore in `WithProducer`?

I don't think it's a huge problem, but this will keep the permit for potentially non-transactional operations

Reply:
Yes, that's what I meant.

Hmm - I agree it's not a huge problem, but it's also not great that `metrics` calls will also potentially block on getting a permit that they don't actually need 🤔

Maybe we should have a separate `WithTransactionalProducer` trait where `apply` has the signature `def apply[A](f: (KafkaByteProducer, Blocking[F], Resource[F, Unit]) => F[A]): F[A]` so that the caller can control how the semaphore is used - what do you think?

(As an aside, I'm not actually convinced that `WithProducer` carries its weight as an abstraction - it would make more sense to have if the underlying producer were pooled or hot-swappable, but it isn't. But as long as we do have that abstraction, I think the lifecycle of the semaphore should be managed in the same place as the lifecycle of the producer.)